UI Modernization: No sites me entry point

momo-ozawa opened the pull request:
Closes #21274 

Design ref: eYeHXEMDbnFptE40xUqZ2T-fi-446_5379

## Description
- Handles the no site scenario for WordPress
    - Updates the no sites screen UI 
    - Adds an entry point to the Me screen so users can log out 
    
    
<img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/30f18f58-ac56-48a6-b37a-5cfaebf7843e" width=200> | <img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/206b934f-27c8-4363-88ac-0e84c94c1772" width=200>
-- | --


## How to test

### WP
- Login to a site with no sites
- Tap "Add new site", verify JP overlay is displayed
- Tap "Account and settings", verify you can logout

iPad | iPhone
-- | --
![Simulator Screenshot - iPad Air (5th generation) - 2023-08-31 at 10 07 54](https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/e791e3d3-1186-407c-a1e7-32a4c85fd1e6) | <img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/0e782778-2af1-49bc-b835-943910c0e602" width=300>

### JP
- Login to a site with no sites
- Tap "Add new site", verify you can add a new site
- Verify "Account and settings" isn't displayed

iPad | iPhone 
-- | --
![Simulator Screenshot - iPad Air (5th generation) - 2023-08-30 at 18 15 38](https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/12a1685d-ea72-47e0-a299-23a6c90e9ef0) | <img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/2f5595c3-b29c-49d4-a259-a761f3f2d041" width=300>


## Regression Notes
1. Potential unintended areas of impact
No sites scenario for both WP and JP

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested manually

3. What automated tests I added (or what prevented me from doing so)
n/a

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

Commit:
Add no sites view

Commit:
add account and settings button

kean commented:
Button color should adjust for Dark mode.

<img width="360" alt="Screenshot 2023-08-29 at 8 25 06 AM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/928e927e-c352-4a3d-82fa-457d51e78016">

kean commented on WordPress/Classes/ViewRelated/Blog/My Site/NoSitesView.swift:
```diff
@@ -0,0 +1,126 @@
+
+    var body: some View {
+        ZStack(alignment: .bottom) {
+            Color.init(uiColor: .systemGroupedBackground)
```
- The `.init` is redundant.
- (In Slack) discussed a different way to change the background

kean commented on WordPress/Classes/ViewRelated/Blog/My Site/NoSitesView.swift:
```diff
@@ -0,0 +1,126 @@
+            .padding(.horizontal, 32)
+
+            if viewModel.isShowingAccountAndSettings {
+                Group {
```
`Group` doesn't affect the layout and is typically used to apply the same modifier to multiple views:

```swift
Group {
    Text("SwiftUI")
    Text("Combine")
    Text("Swift System")
}
.font(.headline)
```

In this case, it applies padding to both the spacer and the accountAndSettingsButton.

The spacer is redundant because the `ZStack` alignment is already set to `.bottom`.

momo-ozawa replied:
Good to know 🙇‍♀️

kean commented on WordPress/Classes/ViewRelated/Blog/My Site/NoSitesView.swift:
```diff
@@ -0,0 +1,126 @@
+            Color.init(uiColor: .systemGroupedBackground)
+                .edgesIgnoringSafeArea(.all)
+
+            VStack(alignment: .leading, spacing: 16) {
```
(nit) There are two subviews in `NoSitesView`: the view in the middle and `accountAndSettingsButton`. I'd suggest extracting the code that creates the view in the middle to a separate property – I think it should be small enough to fit in one property/function. The body will then manage the overall layout of these two views.

Btw, you can also add `@ViewBuilder` to properties or functions if you want to create them using result builders. But if they have a single return statement, it's not necessary.

momo-ozawa replied:
Thanks for the tips!

kean commented on WordPress/Classes/ViewRelated/Blog/My Site/NoSitesView.swift:
```diff
@@ -0,0 +1,126 @@
+}
+
+final class NoSitesViewConfiguration {
+    weak var delegate: NoSitesViewDelegate?
```
(nit) I suggest passing the delegate directly without `NoSitesViewConfiguration`.

kean commented on WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift:
```diff
@@ -686,15 +682,25 @@ class MySiteViewController: UIViewController, NoResultsViewHost, UIScrollViewDel
+
+    func didTapAccountAndSettingsButton() {
+        let meViewController = MeViewController()
+        present(meViewController, animated: true)
```
It seems to be missing a navigation controller, but even with one it seems to work kind of strange:

https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/1b0a470c-2df8-420a-8f3e-ed6692826e8f

momo-ozawa replied:
> but even with one it seems to work kind of strange

It's because I was presenting the vc - I've changed it to push the vc instead 

https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/df56555e-64a4-4f1f-b881-193f2d40ac83

kean commented on WordPress/Classes/ViewRelated/Blog/My Site/NoSitesView.swift:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    var gravatarIcon: some View {
+        let placeholderImage = Image(uiImage: .gridicon(.userCircle, size: CGSize(width: 40, height: 40)))
```
(nit) I'd make it a `var` to avoid creating it when it's not needed.

momo-ozawa replied:
I opted to make the Image in the default case directly

kean commented on WordPress/Classes/ViewRelated/Blog/My Site/NoSitesView.swift:
```diff
@@ -0,0 +1,126 @@
+        return AsyncImage(url: viewModel.gravatarURL) { phase in
+            switch phase {
+            case .success(let image):
+                image.resizable()
```
It will stretch to fill the available space if the `aspectRatio` is not specified.

kean commented on WordPress/Classes/ViewRelated/Blog/My Site/NoSitesView.swift:
```diff
@@ -0,0 +1,126 @@
+        .frame(maxWidth: .infinity, maxHeight: .infinity)
+    }
+
+    var textStackView: some View {
```
Please make these `private`.

Commit:
remove scroll edge appearance

no longer applicable since we’re hiding the nav bar

Commit:
extend view under nav bar

this is to prevent the navigation bar from appearing as transparent when transitioning back to My Site

Commit:
show alert when add new site button is tapped

Commit:
add line limit to account and settings button

Commit:
use a confirmation dialog to display options

https://developer.apple.com/documentation/swiftui/view/confirmationdialog(_:ispresented:titlevisibility:presenting:actions:)-9ibgk

Displays options in a sheet in compact size classes / in a popover in regular size classes

Commit:
fix typo

alpavanoglu commented on WordPress/Classes/ViewRelated/Blog/My Site/NoSitesView.swift:
```diff
@@ -0,0 +1,188 @@
+
+    var body: some View {
+        ZStack(alignment: .bottom) {
+            Color(uiColor: .listBackground)
+                .edgesIgnoringSafeArea(.all)
```
Does this adapt to the theme change? There used to be a problem with `Color:uiColor` init that it lost dynamic nature but rather simply initiated a static color.

momo-ozawa replied:
It does!

<img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/30f18f58-ac56-48a6-b37a-5cfaebf7843e" width=200> | <img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/206b934f-27c8-4363-88ac-0e84c94c1772" width=200>
-- | --

alpavanoglu commented on WordPress/Classes/ViewRelated/Blog/My Site/NoSitesView.swift:
```diff
@@ -0,0 +1,188 @@
+            if viewModel.isShowingAccountAndSettings {
+                makeAccountAndSettingsButton()
+                    .padding(.horizontal, 32)
+                    .padding(.bottom, 20)
```
Is 20 the correct spacing on designs? I think we should be using 8 multipliers.

momo-ozawa replied:
On Figma it looks like its 22 - but I can bump it to 24? cc: @osullivanchris

momo-ozawa replied:
Bumped to 24 [d52a04e](https://github.com/wordpress-mobile/WordPress-iOS/pull/21456/commits/d52a04e5272a5e7756318a28cd4ae76c1aeef922)

alpavanoglu commented on WordPress/Classes/ViewRelated/Blog/My Site/NoSitesView.swift:
```diff
@@ -0,0 +1,188 @@
+        .frame(maxWidth: .infinity, maxHeight: .infinity)
+    }
+
+    private func makeMainView() -> some View {
```
This suggestion/question is quite trivial but just to see what you think about it going forward:

Since in ViewBuilders, we create views as follows:
```Swift
var body: some View {
  firstView
  secondView
}
```

I find it more in line with SwiftUI syntax to use computed properties instead of functions when we can. With functions we need the function call `firstView()` instead of the syntax`firstView`.

Doesn't change anything and it won't work when you need arguments, but since these functions require no arguments I think vars would work nicer here. Wdyt?

momo-ozawa replied:
Done! [765149b](https://github.com/wordpress-mobile/WordPress-iOS/pull/21456/commits/765149bdea96913ebf995f5636d62af9b2325ef4)

alpavanoglu commented on WordPress/Classes/ViewRelated/Blog/My Site/NoSitesView.swift:
```diff
@@ -0,0 +1,188 @@
+    }
+
+    private func makeMainView() -> some View {
+        VStack(alignment: .leading, spacing: 16) {
```
Even though the progress is on hold on Design System, I've added a file named `Length` where you can find virtually all CGFloat options we ought to use for padding, margin, spacing etc. Certainly not required my any means but I wanted to mention its existance:

`Length.Padding.double` would return a `CGFloat` of value 16.0.

momo-ozawa replied:
For this PR I decided not to use `Length` because there's a mixture of 8x and 10x metric values. Going forward I'll keep `Length` in mind - thank you Alp!

alpavanoglu replied:
Just want to note that it could make sense to check with designers if their 10x usage were intentional. Because we had talked about this before with them and afaik they both use the 8x values. I wonder if there's a reason for mixing them together.

momo-ozawa replied:
Updated! https://github.com/wordpress-mobile/WordPress-iOS/pull/21456/commits/768e1889d18274e3c5ff38a9090e4b89db7e83e4

Slack ref: p1693497168975339/1693307309.040859-slack-C04SFL0RP51

alpavanoglu commented on WordPress/Classes/ViewRelated/Blog/My Site/NoSitesView.swift:
```diff
@@ -0,0 +1,188 @@
+    }
+}
+
+struct NoSitesView_Previews: PreviewProvider {
+    static var previews: some View {
+        let configuration = AddNewSiteConfiguration(
+            canCreateWPComSite: true,
+            canAddSelfHostedSite: true,
+            launchSiteCreation: {},
+            launchLoginForSelfHostedSite: {}
+        )
+        NoSitesView(
+            viewModel: NoSitesViewModel(account: nil),
+            addNewSiteConfiguration: configuration
+        )
+    }
+}
```
Does Preview function works in our project? It hasn't been functioning afaik 🤔

momo-ozawa replied:
@kean taught me an easy workaround for this:
- run on Xcode 15 beta 7 (not sure if it works in previous versions)
- comment out willFinishLaunching and didFinishLaunching in WordPressAppDelegate
<img width="2032" alt="Screenshot 2023-08-31 at 10 03 51" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/136d611d-02a7-4c3a-8b23-4a0f3a1fdc86">

alpavanoglu replied:
Interesting, that is actually a pretty useful hack

alpavanoglu commented on WordPress/Classes/ViewRelated/Blog/My Site/NoSitesViewModel.swift:
```diff
@@ -0,0 +1,19 @@
+import Foundation
+
+struct NoSitesViewModel {
```
We could add tests for this VM quite quickly I think.

momo-ozawa replied:
Done! [b2e5be3](https://github.com/wordpress-mobile/WordPress-iOS/pull/21456/commits/b2e5be320b69b45518734dc33dc9d89e460baf8a)

Commit:
use computed variables for views

Commit:
extract accountAndSettingsStackView

Commit:
clarify comment

Commit:
update padding for account and settings button

Commit:
add viewmodel tests

alpavanoglu reviewed (approved):
Thank you for addressing them!

Commit:
add new asset for no site image

Commit:
update padding to use Length

Commit:
add top padding to blog detail header view
